File: iggy.py
#!python
# Copyright (c) 2014, Sven Thiele <sthiele78@gmail.com>
#
# This file is part of iggy.
#
# iggy is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# iggy is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNEOS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with iggy.  If not, see <http://www.gnu.org/licenses/>.
# -*- coding: utf-8 -*-
import sys
import argparse
from pyasp.asp import *
from __iggy__ import query, utils, parsers

if __name__ == '__main__':

  desc = (
    'Iggy confronts biological networks given as interaction graphs with experimental observations '
    'given as signs that represent the concentration changes between two measured states. '
    'Iggy supports the incorporation of uncertain measurements, discovers inconsistencies in data or network, '
    'applies minimal repairs, and predicts the behavior of unmeasured species. '
    'In particular, it distinguishes strong predictions (e.g. increase of a node level) and weak predictions '
    '(e.g., node level increases or remains unchanged).')

  parser = argparse.ArgumentParser(description=desc)
  parser.add_argument("networkfile",
    help="influence graph in SIF format")
  parser.add_argument("observationfile",
    help="observations in bioquali format")

  parser.add_argument('--no_fwd_propagation',
    help="turn forward propagation OFF, default is ON",
    action="store_true")

  parser.add_argument('--no_founded_constraints',
    help="turn constraints OFF that every variation must be founded in an input, default is ON",
    action="store_true")

  parser.add_argument('--elempath',
    help=" a change must be explained by an elementary path from an input.",
    action="store_true")

  parser.add_argument('--depmat',
    help="combines multiple states, a change must be explained by an elementary path from an input.",
    action="store_true")

  parser.add_argument('--mics',
    help="compute minimal inconsistent cores",
    action="store_true")

  parser.add_argument('--autoinputs',
    help="compute possible inputs of the network (nodes with indegree 0)",
    action="store_true")

  parser.add_argument('--scenfit',
    help="compute scenfit of the data, default is mcos",
    action="store_true")

  parser.add_argument('--show_labelings',type=int, default=-1,
    help="number of labelings to print, default is OFF, 0=all")

  parser.add_argument('--show_predictions',
    help="show predictions",
    action="store_true")


  args = parser.parse_args()

  net_string = args.networkfile
  obs_string = args.observationfile

  FP  = not (args.no_fwd_propagation)
  FC  = not (args.no_founded_constraints)
  EP  = args.elempath
  DM  = args.depmat

  print('_____________________________________________________________________\n')
  if DM :
    print(' + DepMat combines multiple states.')
    print(' + An elementary path from an input must exist to explain changes.')
    OS = False
    EP = True
    FP = True
    FC = True
    
  else :
    print(' + All observed changes must be explained by an predecessor.')
    OS = True
    if FP : print(' + 0-change must be explained.')
    if FC : print(' + All observed changes must be explained by an input.')
    if EP : print(' + An elementary path from an input must exist to explain changes.')

  print('_____________________________________________________________________')

  #if (not args.scenfit) and EP :
    #print('\nMCoS and elementary path / DepMat do not work well together.'
          #'Please use --scenfit !')
    #exit()


  print('\nReading network',net_string, '... ',end='')
  net = parsers.readSIFGraph(net_string)
  print('done.')

  # gather some stats on the network
  activations = set()
  inhibitions = set()
  nodes       = set()
  for a in net:
    if a.pred() == 'obs_elabel' :
      if a.arg(2) == '1'  : activations.add((a.arg(0),a.arg(1)))
      if a.arg(2) == '-1' : inhibitions.add((a.arg(0),a.arg(1)))
    if a.pred() == 'vertex' : nodes.add(a.arg(0))
  unspecified = activations & inhibitions

  print('\nNetwork stats:')
  print("         Nodes =", len(nodes))
  print("   Activations =", len(activations))
  print("   Inhibitions =", len(inhibitions))
  print("          Dual =", len(unspecified))


  print('\nReading observations',obs_string, '... ',end='')
  mu = parsers.readProfile(obs_string)
  print('done.')


  print('\nChecking observations',obs_string, '... ',end='')
  contradictions = query.get_contradictory_obs(mu)
  print('done.')
  if len(contradictions) == 0 : print('\nObservations are OK!')
  else:
    print('\nContradictory observations found. Please correct manually!')
    for c in contradictions : print ('  ',c)
    utils.clean_up()
    exit()

  # gather some stats on the observations  
  plus     = set()
  zero     = set()
  minus    = set()
  notminus = set()
  notplus  = set()
  inputs   = set()
  for a in mu:
    if a.pred() == 'obs_vlabel' :
      if a.arg(2) == '1'          : plus.add(a.arg(1))
      if a.arg(2) == '0'          : zero.add(a.arg(1))
      if a.arg(2) == '-1'         : minus.add(a.arg(1))
      if a.arg(2) == 'notMinus'   : notminus.add(a.arg(1))
      if a.arg(2) == 'notPlus'    : notplus.add(a.arg(1))
    if a.pred() == 'input'      : inputs.add(a.arg(1))

  unobserved   = nodes -(plus|minus|zero|notplus| notminus)
  not_in_model = (plus|minus|notplus|zero|notminus)-nodes

  print("              inputs =", len(inputs&nodes))
  print("          observed + =", len(plus&nodes))
  print("          observed - =", len(minus&nodes))
  print("          observed 0 =", len(zero&nodes))
  print("    observed notPlus =", len(notplus&nodes))
  print("   observed notMinus =", len(notminus&nodes))
  print("          unobserved =", len(unobserved))
  print("        not in model =", len(not_in_model))


  if args.autoinputs :
    print('\nComputing input nodes ... ',end='')
    inputs = query.guess_inputs(net)
    net    = TermSet(net.union(inputs))
    print('done.')
    print("\nNumber of inputs =", len(inputs))

  net_with_data = TermSet(net.union(mu))


  if args.scenfit :
    print('\nComputing scenfit of network and data ... ',end='')
    scenfit = query.get_scenfit(net_with_data, OS, FP, FC, EP)
    print('done.')
    if scenfit == 0 : print("\nThe network and data are consistent: scenfit = 0.")
    else:
      print("\nThe network and data are inconsistent: scenfit = ",str(scenfit),'.',sep='')

      if args.mics:
        print('\nComputing minimal inconsistent cores (mic\'s) ... ',end='')
        sys.stdout.flush()
        mics = query.get_minimal_inconsistent_cores(net_with_data, OS, FP, FC, EP)
        print('done.')
        count  = 1
        oldmic = 0
        for mic in mics:
          if oldmic != mic:
            print('mic ',str(count),':',sep='')
            utils.print_mic(mic.to_list(),net.to_list(),mu.to_list())
            count += 1
            oldmic = mic


    if args.show_labelings >= 0 :
      print('\nCompute scenfit labelings ... ',end='')
      labelings = query.get_scenfit_labelings(net_with_data, args.show_labelings, OS, FP, FC, EP)
      print('done.')
      count=0
      for l in labelings :
        count+=1
        print('\nLabeling ',str(count),':',sep='')
        utils.print_labeling(l)
        print('\n   Repairs:')
        utils.print_repairs(l)

    if args.show_predictions :
      print('\nCompute predictions under scenfit ... ',end='')
      predictions = query.get_predictions_under_scenfit(net_with_data, OS, FP, FC, EP)
      print('done.')
      print('\nPredictions:')
      utils.print_predictions(predictions)


  if not args.scenfit :
    print('\nComputing mcos of network and data ... ',end='')
    mcos = query.get_mcos(net_with_data, OS, FP, FC, EP)
    print('done.')
    if mcos == 0 : print("\nThe network and data are consistent: mcos = 0.")
    else: 
      print("\nThe network and data are inconsistent: mcos = ",str(mcos),'.',sep='')

      if args.mics:
        print('\nComputing minimal inconsistent cores (mic\'s) ... ',end='')
        sys.stdout.flush()
        mics = query.get_minimal_inconsistent_cores(net_with_data, OS, FP, FC, EP)
        print('done.')
        count  = 1
        oldmic = 0
        for mic in mics:
          if oldmic != mic:
            print('mic ',str(count),':',sep='')
            utils.print_mic(mic.to_list(),net.to_list(),mu.to_list())
            count += 1
            oldmic = mic

    if args.show_labelings >= 0 :
      print('\nCompute mcos labelings ... ',end='')
      labelings = query.get_mcos_labelings(net_with_data, args.show_labelings, OS, FP, FC, EP)
      print('done.')
      count = 0
      for l in labelings :
        count += 1
        print('\nLabeling ',str(count),':',sep='')
        utils.print_labeling(l)
        print('\n   Repairs:')
        utils.print_repairs(l)

    if args.show_predictions :
      print('\nCompute predictions under mcos ... ',end='')
      predictions = query.get_predictions_under_mcos(net_with_data, OS, FP, FC, EP)
      print('done.')
      print('\nPredictions:')
      utils.print_predictions(predictions)


  utils.clean_up()

File: src/bin/iggy.rs
use anyhow::{anyhow, Context, Result};
use clap::Parser;
use clingo::FactBase;
use log::{error, info, warn};
use std::{fs::File, io::Write, path::PathBuf};
use stderrlog;

use iggy::{cif_parser::Graph, misc::*, CheckResult::Inconsistent, *};

/// Iggy confronts interaction graph models with observations of (signed) changes between two measured states
/// (including uncertain observations).
/// Iggy discovers inconsistencies in networks or data, applies minimal repairs, and
/// predicts the behavior for the unmeasured species. It distinguishes strong predictions (e.g. increase in a
/// node) and weak predictions (e.g., the value of a node increases or remains unchanged).
#[derive(Parser, Debug)]
#[clap(version, author)]
struct Opt {
    /// Influence graph in CIF format
    #[clap(short = 'n', long = "network", value_name = "FILE", parse(from_os_str))]
    network_file: PathBuf,

    /// Observations in bioquali format
    #[clap(
        short = 'o',
        long = "observations",
        value_name = "FILE",
        parse(from_os_str)
    )]
    observations_file: Option<PathBuf>,

    /// Disable forward propagation constraints
    #[clap(long, conflicts_with = "depmat")]
    fwd_propagation_off: bool,

    /// Disable foundedness constraints
    #[clap(long, conflicts_with = "depmat")]
    founded_constraints_off: bool,

    /// Every change must be explained by an elementary path from an input
    #[clap(long)]
    elempath: bool,

    /// Combine multiple states, a change must be explained by an elementary path from an input
    #[clap(long)]
    depmat: bool,

    /// Compute minimal inconsistent cores
    #[clap(long)]
    mics: bool,

    /// Declare nodes with indegree 0 as inputs
    #[clap(short = 'a', long)]
    auto_inputs: bool,

    /// Compute scenfit of the data, default is mcos
    #[clap(long)]
    scenfit: bool,

    /// Show N labelings, default is OFF, 0=all
    #[clap(short = 'l', long = "show-labelings", value_name = "N")]
    max_labelings: Option<u32>,

    /// Show predictions
    #[clap(short = 'p', long)]
    show_predictions: bool,

    /// Multithreading
    #[clap(short = 't', long, value_name = "N", default_value_t = 1)]
    threads: u8,

    /// Print JSON output
    #[clap(long)]
    json: bool,
}

fn main() {
    stderrlog::new()
        .module(module_path!())
        .verbosity(2)
        .init()
        .unwrap();
    if let Err(err) = run() {
        println!("\n## Error\n\n{err}");
        error!("{:?}", err);
        std::process::exit(1);
    }
}

fn run() -> Result<()> {
    let opt = Opt::parse();
    let mut out = std::io::stdout();
    if opt.json {
        writeln!(out, "{{")?;
    } else {
        writeln!(out, "# Iggy Report")?;
    }

    let graph = read_network(&opt, &mut out)?;
    let graph_facts = graph.to_facts();

    let profile = {
        if let Some(observationfile) = &opt.observations_file {
            info!("Reading observations ...");
            if opt.json {
                writeln!(out, ",\"Observations file\":{:?}", observationfile)?;
            } else {
                writeln!(
                    out,
                    "\n## Observations\n\n- Filename: {}",
                    observationfile.display()
                )?;
            }
            let f = File::open(&observationfile)
                .context(format!("unable to open '{}'", observationfile.display()))?;
            let pprofile = profile_parser::read(&f, "x1")
                .context(format!("unable to parse '{}'", observationfile.display()))?;

            let observations_statistics = observations_statistics(&pprofile, &graph);
            if opt.json {
                let serialized = serde_json::to_string(&observations_statistics)?;
                writeln!(out, ",\"Observations Statistics\":{serialized}")?;
            } else {
                write_observation_statistics(&mut out, &observations_statistics)?;
            }
            let profile = pprofile.to_facts();

            info!("Checking observations ...");
            if let Inconsistent(reasons) = check_observations(&profile)? {
                warn!("Contradictory observations. Please correct them!");
                Err(anyhow!(
                    "\nInconsistent observations in {}\n- {}",
                    observationfile.display(),
                    reasons.join("\n- ")
                ))?;
            }
            profile
        } else {
            warn!("Empty observation data.");
            FactBase::new()
        }
    };

    let setting = get_setting(&opt);
    if opt.json {
        write_setting_json(&mut out, &setting)?;
    } else {
        write_setting_md(&mut out, &setting)?;
    }

    let auto_inputs = compute_auto_inputs(&opt, &graph_facts, &mut out)?;

    if opt.scenfit {
        info!("Computing scenfit of network and data ...");
        let scenfit = get_scenfit(&graph_facts, &profile, &auto_inputs, &setting, opt.threads)?;

        if scenfit == 0 {
            info!("The network and data are consistent");
            if opt.json {
                writeln!(out, ",\"scenfit\":0")?;
            } else {
                writeln!(out, "\n## Consistency score\n\n- scenfit: 0")?;
            }
        } else {
            info!("The network and data are inconsistent");
            if opt.json {
                writeln!(out, ",\"scenfit\":{scenfit}")?;
            } else {
                writeln!(out, "\n## Consistency score\n\n- scenfit: {scenfit}")?;
            }
            if opt.mics {
                let mics = get_minimal_inconsistent_cores(
                    &graph_facts,
                    &profile,
                    &auto_inputs,
                    &setting,
                    opt.threads,
                )?;
                if opt.json {
                    write_json_mics(&mut out, mics)?;
                } else {
                    write_mics(&mut out, mics)?;
                }
            }
        }
        if let Some(max_labelings) = opt.max_labelings {
            let l = get_scenfit_labelings(
                &graph_facts,
                &profile,
                &auto_inputs,
                max_labelings,
                &setting,
                opt.threads,
            )?;
            if opt.json {
                write_json_labelings(&mut out, l)?;
            } else {
                write_labelings(&mut out, l)?;
            }
        }
        if opt.show_predictions {
            info!("Compute predictions ...");
            let predictions =
                get_predictions_under_scenfit(&graph_facts, &profile, &auto_inputs, &setting)?;

            if opt.json {
                write_json_predictions(&mut out, &predictions)?;
            } else {
                write_predictions(&mut out, &predictions)?;
            }
        }
    } else {
        info!("Computing mcos of network and data ...");
        let mcos = get_mcos(&graph_facts, &profile, &auto_inputs, &setting, opt.threads)?;
        if mcos == 0 {
            info!("The network and data are consistent");
            if opt.json {
                writeln!(out, ",\"mcos\":0")?;
            } else {
                writeln!(out, "\n## Consistency score\n\n- mcos: 0")?;
            }
        } else {
            info!("The network and data are inconsistent");
            if opt.json {
                writeln!(out, ",\"mcos\":{mcos}")?;
            } else {
                writeln!(out, "\n## Consistency score\n\n- mcos: {mcos}")?;
            }
            if opt.mics {
                let mics = get_minimal_inconsistent_cores(
                    &graph_facts,
                    &profile,
                    &auto_inputs,
                    &setting,
                    opt.threads,
                )?;
                if opt.json {
                    write_json_mics(&mut out, mics)?;
                } else {
                    write_mics(&mut out, mics)?;
                }
            }
        }
        if let Some(max_labelings) = opt.max_labelings {
            let l = get_mcos_labelings(
                &graph_facts,
                &profile,
                &auto_inputs,
                max_labelings,
                &setting,
                opt.threads,
            )?;
            if opt.json {
                write_json_labelings(&mut out, l)?;
            } else {
                write_labelings(&mut out, l)?;
            }
        }
        if opt.show_predictions {
            info!("Compute predictions ...");
            let predictions =
                get_predictions_under_mcos(&graph_facts, &profile, &auto_inputs, &setting)?;
            if opt.json {
                write_json_predictions(&mut out, &predictions)?;
            } else {
                write_predictions(&mut out, &predictions)?;
            }
        }
    }
    if opt.json {
        print!("}}");
    }
    Ok(())
}

fn compute_auto_inputs(
    opt: &Opt,
    graph_facts: &FactBase,
    out: &mut std::io::Stdout,
) -> Result<FactBase, anyhow::Error> {
    let auto_inputs = {
        if opt.auto_inputs {
            info!("Computing input nodes ...");
            let inputs = get_auto_inputs(graph_facts)?;
            let node_ids = get_node_ids_from_inputs(&inputs)?;
            if opt.json {
                write_auto_inputs_json(out, &node_ids)?;
            } else {
                write_auto_inputs_md(out, &node_ids)?
            }
            inputs
        } else {
            FactBase::new()
        }
    };
    Ok(auto_inputs)
}

fn get_setting(opt: &Opt) -> Setting {
    let setting = if opt.depmat {
        Setting {
            os: false,
            ep: true,
            fp: true,
            fc: true,
        }
    } else {
        Setting {
            os: true,
            ep: opt.elempath,
            fp: !opt.fwd_propagation_off,
            fc: !opt.founded_constraints_off,
        }
    };
    setting
}

fn read_network(opt: &Opt, mut out: impl Write) -> Result<Graph> {
    info!("Reading network model ...");
    if opt.json {
        writeln!(out, "\"Network file\":{:?}", opt.network_file)?;
    } else {
        writeln!(
            out,
            "\n## Network\n\n- Filename: {}",
            opt.network_file.display()
        )?;
    }
    let f = File::open(&opt.network_file)
        .context(format!("unable to open '{}'", opt.network_file.display()))?;
    let graph = cif_parser::read(&f)
        .context(format!("unable to parse '{}'", opt.network_file.display()))?;
    let network_statistics = graph.statistics();
    if opt.json {
        let serialized = serde_json::to_string(&network_statistics)?;
        writeln!(out, ",\"Network statistics\":{serialized}")?;
    } else {
        writeln!(out, "{network_statistics}")?;
    }
    Ok(graph)
}
